File: paleo_workbench/mapping/topology_checker.py
"""拓扑检查器宿主面（拓扑编辑迁移 M4 §5）。

桥 ``run_geometry_checks`` / ``fix_geometry_error`` 的结果 + 双豁免
（忽略列表、缝隙白名单）+ 保存门禁过滤。工区余量由桥自制规则返回。
"""
from __future__ import annotations

import json
from collections.abc import Iterable, Mapping
from datetime import datetime, timezone

__all__ = ["TopologyChecker", "ignore_key"]


def ignore_key(error: Mapping[str, object]) -> tuple[str, str, str, str]:
    """忽略身份：规则 + 层 + 要素 + 对方要素（不依赖一次运行的临时 id）。"""
    return (
        str(error.get("rule") or ""),
        str(error.get("layer_id") or ""),
        str(error.get("feature_id") or ""),
        str(error.get("other_feature_id") or ""),
    )


class TopologyChecker:
    """上次检查结果 + 豁免 + 桥运行/修复。"""

    def __init__(self) -> None:
        self.last_errors: list[dict[str, object]] = []
        self.last_run_at: str | None = None
        self.workspace: dict | None = None
        self._ignored: dict[tuple[str, str, str, str], str] = {}
        self.allowed_gaps: list[dict] = []

    # -- 豁免 ----------------------------------------------------------------

    def ignore(self, error: Mapping[str, object], *, reason: str = "") -> None:
        self._ignored[ignore_key(error)] = str(reason or "")

    def restore(self, error: Mapping[str, object]) -> None:
        self._ignored.pop(ignore_key(error), None)

    def is_ignored(self, error: Mapping[str, object]) -> bool:
        return ignore_key(error) in self._ignored

    def ignored_keys(self) -> set[tuple[str, str, str, str]]:
        return set(self._ignored)

    def add_allowed_gap(self, geometry: Mapping[str, object]) -> None:
        self.allowed_gaps.append(dict(geometry))

    def blocking_errors(self, errors: Iterable[Mapping[str, object]] | None = None
                        ) -> list[dict[str, object]]:
        source = list(errors) if errors is not None else list(self.last_errors)
        return [dict(error) for error in source if not self.is_ignored(error)]

    def persist(self) -> dict[str, object]:
        return {
            "ignored": [
                {
                    "rule": key[0],
                    "layer_id": key[1],
                    "feature_id": key[2],
                    "other_feature_id": key[3],
                    "reason": reason,
                }
                for key, reason in self._ignored.items()
            ],
            "allowed_gaps": list(self.allowed_gaps),
            "workspace": dict(self.workspace) if self.workspace else None,
        }

    def restore_state(self, snapshot: Mapping[str, object] | None) -> None:
        self._ignored.clear()
        self.allowed_gaps = []
        self.workspace = None
        if not snapshot:
            return
        for entry in snapshot.get("ignored") or ():
            if not isinstance(entry, Mapping):
                continue
            self._ignored[ignore_key(entry)] = str(entry.get("reason") or "")
        for geom in snapshot.get("allowed_gaps") or ():
            if isinstance(geom, Mapping):
                self.allowed_gaps.append(dict(geom))
        workspace = snapshot.get("workspace")
        if isinstance(workspace, Mapping):
            self.workspace = dict(workspace)

    # -- 运行 ----------------------------------------------------------------

    def run_for_commit(self, stack, canvas, layer_ids) -> list[dict[str, object]]:
        """保存前自动全量检查；返回未忽略错误（门禁输入）。"""
        self.run(stack, canvas, layer_ids)
        return self.blocking_errors()

    def run(self, stack, canvas, layer_ids, *, extra_config=None) -> list[dict[str, object]]:
        config = {
            "layer_ids": [str(layer_id) for layer_id in layer_ids],
            "rules": ["overlap", "gap", "is_valid", "workspace_remainder"],
            "precision": 8,
        }
        if self.workspace:
            config["workspace"] = self.workspace
        if self.allowed_gaps:
            config["allowed_gaps"] = {
                "type": "FeatureCollection",
                "features": [
                    {"type": "Feature", "geometry": geom, "properties": {}}
                    for geom in self.allowed_gaps
                ],
            }
        if extra_config:
            config.update(extra_config)
        raw = stack.run_geometry_checks(canvas, json.dumps(config))
        payload = raw if isinstance(raw, dict) else json.loads(str(raw or "{}"))
        self.last_errors = [
            dict(error) for error in (payload.get("errors") or [])
            if isinstance(error, Mapping)
        ]
        harvested = payload.get("allowed_gaps")
        if isinstance(harvested, list) and harvested:
            self.allowed_gaps = [
                dict(geom) for geom in harvested if isinstance(geom, Mapping)
            ]
        self.last_run_at = datetime.now(timezone.utc).isoformat()
        return list(self.last_errors)

    def fix(self, stack, canvas, error_id, method: int = 0) -> dict[str, object]:
        raw = stack.fix_geometry_error(canvas, str(error_id), int(method))
        payload = raw if isinstance(raw, dict) else json.loads(str(raw or "{}"))
        if isinstance(payload.get("errors"), list):
            self.last_errors = [dict(e) for e in payload["errors"]
                                if isinstance(e, Mapping)]
        return payload

    def fix_all(self, stack, canvas, error_ids, method: int = 0) -> dict[str, object]:
        raw = stack.fix_geometry_errors(
            canvas, json.dumps([str(i) for i in error_ids]), int(method))
        payload = raw if isinstance(raw, dict) else json.loads(str(raw or "{}"))
        if isinstance(payload.get("errors"), list):
            self.last_errors = [dict(e) for e in payload["errors"]
                                if isinstance(e, Mapping)]
        return payload

File: libs/ui_composite/src/topology_checker.cpp
#include <pwb/ui_composite/topology_checker.hpp>

#include <chrono>
#include <cstdio>
#include <stdexcept>

namespace pwb::ui_composite {
namespace {

std::string str_field(const Json& object, const char* key) {
    if (!object.is_object()) {
        return {};
    }
    auto it = object.find(key);
    if (it == object.end() || it->is_null()) {
        return {};
    }
    if (it->is_string()) {
        return it->get<std::string>();
    }
    if (it->is_number_integer()) {
        return std::to_string(it->get<long long>());
    }
    if (it->is_number()) {
        return std::to_string(it->get<double>());
    }
    if (it->is_boolean()) {
        return it->get<bool>() ? "True" : "False";
    }
    return {};
}

// Raw bridge payload normalization: object passes through, JSON text is
// parsed, anything else becomes {}.
Json parse_payload(const Json& raw) {
    if (raw.is_object()) {
        return raw;
    }
    if (raw.is_string()) {
        try {
            const Json parsed = Json::parse(raw.get<std::string>());
            return parsed.is_object() ? parsed : Json::object();
        } catch (...) {
            return Json::object();
        }
    }
    return Json::object();
}

std::vector<Json> error_list(const Json& payload) {
    std::vector<Json> out;
    auto it = payload.find("errors");
    if (it != payload.end() && it->is_array()) {
        for (const Json& entry : *it) {
            if (entry.is_object()) {
                out.push_back(entry);
            }
        }
    }
    return out;
}

std::string utc_now_iso() {
    const auto now = std::chrono::system_clock::now();
    const std::time_t t = std::chrono::system_clock::to_time_t(now);
    std::tm tm{};
#if defined(_WIN32)
    gmtime_s(&tm, &t);
#else
    gmtime_r(&t, &tm);
#endif
    char buf[32];
    std::snprintf(buf, sizeof(buf), "%04d-%02d-%02dT%02d:%02d:%02d+00:00",
                  tm.tm_year + 1900, tm.tm_mon + 1, tm.tm_mday, tm.tm_hour,
                  tm.tm_min, tm.tm_sec);
    return buf;
}

}  // namespace

IgnoreKey ignore_key(const Json& error) {
    return {str_field(error, "rule"), str_field(error, "layer_id"),
            str_field(error, "feature_id"),
            str_field(error, "other_feature_id")};
}

void TopologyChecker::ignore(const Json& error,
                             const std::string& reason) {
    ignored_[ignore_key(error)] = reason;
}

void TopologyChecker::restore(const Json& error) {
    ignored_.erase(ignore_key(error));
}

bool TopologyChecker::is_ignored(const Json& error) const {
    return ignored_.count(ignore_key(error)) != 0;
}

std::set<IgnoreKey> TopologyChecker::ignored_keys() const {
    std::set<IgnoreKey> out;
    for (const auto& [key, reason] : ignored_) {
        out.insert(key);
    }
    return out;
}

void TopologyChecker::add_allowed_gap(const Json& geometry) {
    allowed_gaps.push_back(geometry);
}

std::vector<Json> TopologyChecker::blocking_errors(
    const std::optional<std::vector<Json>>& errors) const {
    const std::vector<Json>& source =
        errors.has_value() ? *errors : last_errors;
    std::vector<Json> out;
    for (const Json& error : source) {
        if (!is_ignored(error)) {
            out.push_back(error);
        }
    }
    return out;
}

Json TopologyChecker::persist() const {
    Json ignored_list = Json::array();
    for (const auto& [key, reason] : ignored_) {
        ignored_list.push_back({
            {"rule", key[0]},
            {"layer_id", key[1]},
            {"feature_id", key[2]},
            {"other_feature_id", key[3]},
            {"reason", reason},
        });
    }
    return {
        {"ignored", std::move(ignored_list)},
        {"allowed_gaps", allowed_gaps},
        {"workspace",
         workspace.has_value() ? *workspace : Json(nullptr)},
    };
}

void TopologyChecker::restore_state(const Json& snapshot) {
    ignored_.clear();
    allowed_gaps.clear();
    workspace = std::nullopt;
    if (!snapshot.is_object()) {
        return;
    }
    auto ignored_it = snapshot.find("ignored");
    if (ignored_it != snapshot.end() && ignored_it->is_array()) {
        for (const Json& entry : *ignored_it) {
            if (!entry.is_object()) {
                continue;
            }
            ignored_[ignore_key(entry)] = str_field(entry, "reason");
        }
    }
    auto gaps_it = snapshot.find("allowed_gaps");
    if (gaps_it != snapshot.end() && gaps_it->is_array()) {
        for (const Json& geom : *gaps_it) {
            if (geom.is_object()) {
                allowed_gaps.push_back(geom);
            }
        }
    }
    auto workspace_it = snapshot.find("workspace");
    if (workspace_it != snapshot.end() && workspace_it->is_object()) {
        workspace = *workspace_it;
    }
}

std::vector<Json> TopologyChecker::run(
    const std::vector<std::string>& layer_ids, const Json& extra_config) {
    if (!run_fn_) {
        throw std::runtime_error(
            "no geometry-checks bridge installed on TopologyChecker");
    }
    Json config = {
        {"layer_ids", layer_ids},
        {"rules",
         {"overlap", "gap", "is_valid", "workspace_remainder", "dangle"}},
        {"precision", 8},
    };
    if (workspace.has_value()) {
        config["workspace"] = *workspace;
    }
    if (!allowed_gaps.empty()) {
        Json features = Json::array();
        for (const Json& geom : allowed_gaps) {
            features.push_back(
                {{"type", "Feature"},
                 {"geometry", geom},
                 {"properties", Json::object()}});
        }
        config["allowed_gaps"] = {{"type", "FeatureCollection"},
                                  {"features", std::move(features)}};
    }
    if (extra_config.is_object()) {
        for (const auto& [key, value] : extra_config.items()) {
            config[key] = value;
        }
    }
    const Json payload = parse_payload(run_fn_(config));
    last_errors = error_list(payload);
    auto harvested = payload.find("allowed_gaps");
    if (harvested != payload.end() && harvested->is_array() &&
        !harvested->empty()) {
        allowed_gaps.clear();
        for (const Json& geom : *harvested) {
            if (geom.is_object()) {
                allowed_gaps.push_back(geom);
            }
        }
    }
    last_run_at = utc_now_iso();
    return last_errors;
}

std::vector<Json> TopologyChecker::run_for_commit(
    const std::vector<std::string>& layer_ids) {
    run(layer_ids);
    return blocking_errors();
}

Json TopologyChecker::fix(const std::string& error_id, int method) {
    if (!fix_fn_) {
        throw std::runtime_error(
            "no geometry-fix bridge installed on TopologyChecker");
    }
    const Json payload = parse_payload(fix_fn_(error_id, method));
    if (auto it = payload.find("errors");
        it != payload.end() && it->is_array()) {
        last_errors = error_list(payload);
    }
    return payload;
}

Json TopologyChecker::fix_all(const std::vector<std::string>& error_ids,
                              int method) {
    if (!fix_all_fn_) {
        throw std::runtime_error(
            "no geometry-fix bridge installed on TopologyChecker");
    }
    const Json payload = parse_payload(fix_all_fn_(error_ids, method));
    if (auto it = payload.find("errors");
        it != payload.end() && it->is_array()) {
        last_errors = error_list(payload);
    }
    return payload;
}

}  // namespace pwb::ui_composite
